[2단계 - 자동차 경주 리팩터링] 엘리(한해리) 미션 제출합니다.

RIANAEH opened the pull request:
안녕하세요 희봉!!
벌써 step2 리뷰 요청을 보내게되었네요 
시간이 참 빨리 흐르는 것 같아요 😅

이번에는 step2 요구사항을 보고 MVC 패턴에 대한 이해가 부족하다고 생각돼서 이에 대한 학습을 진행한 후 현재 제 프로젝트를 어떻게 하면 MVC 패턴에 부합되도록 리팩터링할지 계속 고민하면서 미션을 진행했어요!!
그러면서 계속해서 “이 객체에서 이 기능을 제공하는 것이 맞을까?”, “너무 다른 객체와 의존관계를 갖는게 아닐까?”를 의심하면서 프로젝트 구조를 바꿔 나갔던겄같아요. 
많은 조언 부탁드립니다 🙇‍♀️

### 프로젝트 구조 변경 사항
- Controller : RacingGame, CarController(X) 
- Model : Car, Winner(X), RacingGame(new)
- View : InputView, OutputView

### CarController가 정말 컨트롤러의 역할을 하는가?
처음에는 CarController의 역할에 대한 의심부터 시작됐어요. MVC를 다시 학습한 후 바라보니 컨트롤러가 아닌 모델의 역할을 하고 있더라구요. 요구사항 그림에서 힌트를 얻어 이를 RacingGame으로 변경했어요. 

### Winner가 정말 필요한가?
이전에 희봉이 Winner 클래스 생성에 대해 칭찬을 해주셨었는데, CarController를 RacingCar로 변경하고 Winner를 살펴보니 Winner에서 RacingCar의 데이터인 cars를 받아와 사용한다는 것을 인지하게 됐어요.
이러한 방식은 getter 사용을 유발하는데, getter의 경우 뷰와의 상호작용을 위한 기능 외에서는 그 사용을 자제해야한다고 생각하기에 아예 RacingCar 자체에서 본인의 데이터인 cars를 사용해 우승자를 가려내는게 더 좋은 방법이지 않을까 하는 생각이 들었어요. 
그리고 아예 우승자를 가려내는 구현 로직을 변경하면 maxPosition 필드 마저 없앨 수 있겠다는 생각이 들면서 Winner를 없애는 방향으로 생각을 굳히게되었어요. 
하지만 이렇게 되면 "우승자 구함"이라는 책임을 RacingCar에 추가적으로 부담시키게 되는 것이기 때문에 이 방법이 정말 더 좋은 방법이 맞는 지는 사실 계속 의문이 들어요 🥲

### 우승자 선정 구현 로직 변경
기존의 우승자 선정을 구현할 때 getter로 position을 받아와 이를 비교하여 가장 큰 position을 찾아내는 방법을 사용했는데, 생각해보니 굳이 getter로 Car 객체의 데이터를 밖으로 끌어봐서 비교를 해야하는지 의문이 들었어요. 그래서 객체의 필드끼리 비교할 수 있는 다른 방법이 없을까 찾아보다 객체 정렬을 커스터마이징할 수 있다는 것을 보고 이를 사용했어요. 
Comparable 인터페이스를 사용해 Car 클래스 내에 비교 메서드를 오버라이딩해서 사용하니 position을 밖으로 빼와서 비교할 필요가 없더라구요. 
하지만 stream x 2와 정렬 + stream이 성능적으로 어떤 차이를 보일지는 아직 모르겠어서 더 학습이 필요할 것 같아요 😅 

### 우승자 선정 테스트
Winner의 역할을 모두 RacingGame으로 이동시키고 나니 테스트 코드 또한 변경이 필요했어요. 하지만 기존에 우승자 선정을 테스트하는 방식으로 RacingGameTest에서 테스트해보려했더니 RacingGame 생성자로 carNames, tryCount를 받는 방식때문에 독립적으로(?) RacingGame 객체를 만들어낼 수가 없어 테스트가 불가능했어요. 
그래서 아예 carNames를 이용해 Car 객체들을 생성하는 로직은 GameController에서 진행하고 RacingGame 생성자에서는 cars, tryCount를 받게 변경해 테스트를 할 수 있게 했어요. 이렇게 변경하고 보니 오히려 RacingCar의 Car에 대한 의존성을 줄이는 효과가 생겨난 것 같아 뭔가 뿌듯했어요 😊

## 궁금증
### 축약은 어디까지해야하는가..
다음의 1번과 2번 코드 중 어떤게 더 좋은 방법인지 고민이 되더라구요. 제가 생각하기에는 1번이 더 가독성이 좋다고 느껴졌지만 2번은 코드가 단축된다는 장점이 있더라구요. 보통 현업에서는 어떤 방법을 사용하나요?
```java
    private void ready() {
        final List<Car> cars = createCars(inputCarNames());
        final int tryCount = inputTryCount();

        racingGame = new RacingGame(cars, tryCount);
    }
```
```java
    private void ready() {
        racingGame = new RacingGame(createCars(inputCarNames(), inputTryCount());
    }
```

### @DisplayName 어노테이션의 위치?
기존에`@Test` 어노테이션을 사용할 때는 그냥 무의식적으로 1번과 같이 사용했는데 `@ParameterizedTest`에 대해 찾아보다보니 2번과 같이 사용하시는 분들도 있더라구요. 둘 중 어느게 컨벤션에 부합하는지 찾기 힘들어서 이렇게 질문 드립니다!!
```java
@ParameterizedTest
@ValueSource(ints = {5, 6, 7, 8, 9})
@DisplayName("자동차 전진 테스트")
void move(int number) {
    car.moveForward(number);
    assertThat(car.getPosition()).isEqualTo(1);
}
```
```java
@DisplayName("자동차 전진 테스트")
@ParameterizedTest
@ValueSource(ints = {5, 6, 7, 8, 9})
void move(int number) {
    car.moveForward(number);
    assertThat(car.getPosition()).isEqualTo(1);
}
```

RIANAEH committed:
refactor(racingcar): CarController를 RacingGame으로 변경

CarController가 사실은 컨드롤러의 역할이 아닌 모델의 역할을 수행하고 있다고 생각되어 이를 RacingGame으로 변경하였습니다.

RIANAEH committed:
refactor(racingcar): Winner의 기능을 RacingGame으로 이동

Winner에서 RacingGame의 데이터를 받아와 사용하는 방식이 좋지 않다고 판단하여 해당 기능을 RacingGame으로 이동시키면서 Winner를 삭제했습니다.
또한 객체 정렬을 이용하도록 우승자 선정 방법을 변경해보았습니다.

RIANAEH committed:
refactor(racingcar): 테스트 케이스 추가 및 메서드명 영문으로 변경

RIANAEH committed:
refactor(racingcar): 생성자 체인 기법 적용

heebong commented on src/main/java/racingcar/model/Car.java:
```diff
@@ -1,22 +1,27 @@
     public Car(final String name) {
-        this.name = name;
-        position = 0;
+        this(name, 0);
```
생성자 체인 👍

heebong commented on src/main/java/racingcar/controller/GameController.java:
```diff
@@ -29,45 +29,52 @@ public void run() {
     }
 
-    public String[] createCarNames() {
+    public String[] inputCarNames() {
```
현재 `private` 함수에서만 사용되는데 `public`이네요. 테스트를 위해 열어둔 것이겠죠?
`CarNamesValidator.validateCarNames()`의 테스트로 검증할 수 있으니 private으로 바꿔도 될 것 같습니다. 😄

heebong commented on src/main/java/racingcar/model/RacingGame.java:
```diff
@@ -0,0 +1,50 @@
+        tryCount--;
+    }
+
+    public List<String> getWinners() {
```
`CarController`를 `RacingGame`같은 객체로 변경한 것 👏 
이 행동이 controller일까, 아니면 또 다른 도메인 객체인걸까 고민이 되죠 😢  새롭게 리팩터링한 엘리에게 박수를 보냅니다.

물론 이렇게 변경된 후 `Winner`를 사용하려니 `get()` 함수를 사용하게 되어 고민이 들 수 있어요. (`get()`을 결과에만 사용하려는 시도 👍 ) 실은 `get()` 함수를 쓰지 않으려면 한가지 방법밖에 없어요. 
`객체에게 메시지를 보내자`
`get()`을 사용해 하려고 했던 행위를, 해당 객체에게 메시지를 보내서 해결할 수 있습니다. 여기서는? `getCars()`인 `List<Car>`에게 메시지를 보내볼 수 있겠네요 😄 

물론 정답이라는건 아닙니다. 일전에 말했다시피 모든건 트레이드 오프에요. 현재 장점이 극대화되는 방법을 선택하는 것이 중요합니다. 기존 코드를 유지한다면 관리 포인트를 적게 가져갈 수 있겠고 새로운 객체를 가져간다면 코드 응집도가 더 높아지겠죠. 요구사항은 언제든 변경될 수 있고, 상황에 맞게 더 좋은 방식으로 리팩터링이 무조건 일어난다는 것만 생각하면 됩니다. (물론 지금은 연습이니까 다양한 시도를 해보는 것도 좋아요 😉 )

RIANAEH replied:
**"현재 장점이 극대화되는 방법을 선택하는 것이 중요하다."** 
매우 마음에 와닿는 말씀인 것 같아요!! 하지만 이러한 선택을 하기위해서는 선택지들이 어떠한 장점을 가지는 지 단점을 가져야하는 지 알아야하겠죠.. 🤣 이런 것을 알아 보는 눈을 기르기 위해 열심히 공부해야겠어요!! 👍

이 리뷰를 받고 "객체에게 메시지를 보내자"의 의미를 생각해보며 이런 저런 서치를 하고, 제 코드를 보던 와중 궁금한 부분이 생겼어요!!
제가 현재 `RacingGame`에서 두개의 getter 메서드를 사용하고 있더라구요. 
1. `getWinner` 메서드에서 `cars.get(FIRST_WINNER_INDEX)` (이것도 getter로 보는게 맞..겠..죠..?ㅜㅜ)
이 경우 `RacingGame` 본인이 가지고 있는 필드 cars에 대한 접근이기 때문에 가능하다고 생각했는데요. 만약 여기서 리펙터링이 진행된다면 `Cars`와 같은 또 다른 모델 클래스를 만들어야하는건가요? 그런데 그렇게 되면 `Cars`도 똑같이 본인이 가지고 있는 cars에 대한 접근이 필요해지게 되는데.. 그럼 결국 클래스만 나뉠뿐 그 과정은 똑같다고 생각해서요. 혹시 생각을 어떤식으로 접근해야하는 지 조언해주실 수 있을까요?? 🥲
2. `getCars` 메서드
이 메소드의 경우 OutputVIew에서만 사용하는 getter 인데 이런 경우에도 리펙터링 대상이 되나요??

heebong replied:
RacingGame이 winner를 결정하는 역할까지 하는게 고민인 것 같아서 드린 이야기랍니다. 현재 winner의 결정 방식은 다음과 같죠.
1. `List<Car>`를 position 기준으로 정렬한다.
2. 정렬된 `List<Car>`에서 첫 번째 Car를 꺼낸다.
3. `List<Car>` 안에서 꺼낸 Car와 position이 같은 Car의 이름을 만든다.

만약 메시지를 보낸다면 다음처럼 변할 수 있어요
1. `List<Car>`한테 position 기준으로 첫 번째 Car 달라고 하기
2. `List<Car>`한테 특정 Car와 position이 같은 Car들의 이름을 달라고 하기

하지만 Collection 인터페이스에 우리가 원하는 함수를 추가할 순 없어요. 그래서 List<Car>를 감싼 Cars를 만들어서 함수들을 만들 수 있답니다. `3. List<Car> 안에서 꺼낸 Car와 position이 같은 Car의 이름을 만든다.`와 `2. List<Car>한테 특정 Car와 position이 같은 Car들의 이름을 달라고 하기`의 차이가 느껴지나요? 이 구현이 정답이라는게 아니고, 꺼내서 사용하는 것과 메시지를 보내는 것의 차이를 설명하려고 예시로 든겁니다. (뉘앙스 차이가..잘 전달 되었는지 모르겠네요.. 🥲 ) 엘리는 기본적으로 메시지를 잘 보내고 있기 때문에 앞으로 과제를 해가면서 더 잘 생각해보면 좋겠네요 😄 
참고로 `List<Car>`를 랩핑한 `Cars`가 `List<Car>`에 접근하는걸 걱정하지 않아도 된답니다. `Cars`의 목적은 `List<Car>`를 래핑한것이라 RacingGame과 목적이 다르기 때문에 접근해도 괜찮습니다.

그리고 OutputVIew에서만 사용하는 getter는 리펙터링 대상이 아닙니다! Car가 View로 직접 넘어가기 때문에 당연히 누군가는 Car를 넘겨주어야겠죠? `꺼내서 사용하기보다는 메시지를 보내자`의 관점으로 바라보면 `getter`의 사용이 정리될거에요. 😄

heebong commented on src/main/java/racingcar/model/RacingGame.java:
```diff
@@ -0,0 +1,50 @@
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+import racingcar.utils.RandomNumberGenerator;
```
사용하지 않는 import 제거 👀

RIANAEH replied:
혹시 어떤 부분인지 알 수 있을까요? 
제 IDE에서는 모두 사용하는 걸로 표시돼있어서요 🤔
![image](https://user-images.githubusercontent.com/45311765/154407236-2e4429b2-8ed8-442f-bab6-0371bbfcb8a7.png)

heebong replied:
제가 착각했네요 😅  모두 사용되는 것 맞습니다. 대화는 닫을게요~!

heebong commented on src/main/java/racingcar/model/RacingGame.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    public List<String> getWinners() {
+        Collections.sort(cars);
```
`stream+stream` vs `sorting+stream` 두 코드간의 성능차이를 알고싶다면 간단하게 비교할 수 있는 방법이 있어요. 함수가 시작할 때 시간을 찍고, 끝날때 시간을 찍어서 둘의 간격을 확인하는 방법입니다. 두 개 중 어떤게 더 빠를지 궁금하다면 확인해 볼 수 있어요 😄 

hint : `System.currentTimeMillis()`로 시간 재기, 반복문으로 List<Car> 미리 만들기
간단하게 테스트코드에서 실행해봐요. 확인하는 용도로 커밋은 하지 않습니다 😇

RIANAEH replied:
희봉!! 성능 테스트를 하다보니 궁금한 점들이 생겨났어요 🤣

아래 두 메서드를 비교하였습니다. 
```java
void getWinners1(final List<Car> cars) {
    setMaxPosition(cars);

    cars.stream()
            .filter(car -> car.getPosition() == maxPosition)
            .map(Car::getCarName)
            .collect(Collectors.toList());
}

void setMaxPosition(List<Car> cars) {
    maxPosition = cars.stream()
            .mapToInt(Car::getPosition)
            .max()
            .orElse(0);
}

void getWinners2(List<Car> cars) {
    Collections.sort(cars);

    cars.stream()
            .filter(car -> car.compareTo(cars.get(FIRST_WINNER_INDEX)) == EQUAL_CODE)
            .map(Car::getCarName)
            .collect(Collectors.toList());
}
```

## 반복문으로 테스트를 여러번 실행했을 때와 하나씩 테스트 해봤을 때 왜 각각의 테스트의 성능이 차이가 나는가?
각각 `Car 객체` 10, 100, 1000, 10000, 100000개를 가진 cars에 대해 총 3가지 방법으로 테스트를 해보았어요. 
1. 일일이 테스트
```java
@Test
void comparePerformance() {
    List<Car> cars = new ArrayList<>();
    for (int i = 0; i < 100000; i++) {
        cars.add(new Car("car" + i, i));
    }

    long start = System.currentTimeMillis();
    getWinners1(cars);
    long end = System.currentTimeMillis();

    System.out.println("1번 실행시간 : " + (end - start) / 1000.0 + "초");

    start = System.currentTimeMillis();
    getWinners2(cars);
    end = System.currentTimeMillis();

    System.out.println("2번 실행시간 : " + (end - start) / 1000.0 + "초");

    System.out.println();
}
````
2. `@ParameterizedTest` 이용
```java
@ParameterizedTest
@ValueSource(ints = {10, 100, 1000, 10000, 100000})
void comparePerformance(int count) {
    List<Car> cars = new ArrayList<>();
    for (int i = 0; i < count; i++) {
        cars.add(new Car("car" + i, i));
    }

    long start = System.currentTimeMillis();
    getWinners1(cars);
    long end = System.currentTimeMillis();

    System.out.println("1번 실행시간 : " + (end - start) / 1000.0 + "초");

    start = System.currentTimeMillis();
    getWinners2(cars);
    end = System.currentTimeMillis();

    System.out.println("2번 실행시간 : " + (end - start) / 1000.0 + "초");
}
```
3. 반복문 이용
```java
@Test
void comparePerformance() {
    int[] counts = {10, 100, 1000, 10000, 100000};

    for (int count : counts) {
        measurePerformance(count);
    }
}

void measurePerformance(int count) {
    List<Car> cars = new ArrayList<>();
    for (int i = 0; i < count; i++) {
        cars.add(new Car("car" + i, i));
    }

    long start = System.currentTimeMillis();
    getWinners1(cars);
    long end = System.currentTimeMillis();

    System.out.println("1번 실행시간 : " + (end - start) / 1000.0 + "초");

    start = System.currentTimeMillis();
    getWinners2(cars);
    end = System.currentTimeMillis();

    System.out.println("2번 실행시간 : " + (end - start) / 1000.0 + "초");

    System.out.println();
}
```

처음에는 일일이 테스트를 해보다가 `@ParameterizedTest`를 사용해서 테스트해볼 수 있겠다고 생각해서 이를 사용해봤어요. 하지만 그랬더니 결과가 다음과 같이 나오더라구요. 
![image](https://user-images.githubusercontent.com/45311765/154497026-9c60871d-fbd1-4760-b825-beeaeb52adc7.png)
이 결과는 일일이 테스트해봤을 때와는 완전히 달랐어요. 그때는 테스트값의 개수가 많아질 수록 실행시간이 늘어났는데(이게 당연한건데..) 여기서는 결과가 이상하더라구요. 
저는 지금까지 `@ParameterizedTest`의 원리가 단순히 같은 메서드에 다른 파라미터값을 거내면서 반복을 하는것이라고 생각했어서 궁금한 마음에 단순히 `for문`을 사용해 테스트를 해보았더니 `@ParameterizedTest`와 같은 결과가 나오더라구요... 이랬을 경우 결과가 이렇게 나오는 이유를 알 수 있을까요..? 뭔가 감이 안잡힙니다... 🥲

## 실제 현업에서 이러한 경우 어떠한 기준으로 코드 구현 방법을 결정하는가?
일일이 테스트를 해봤을 경우 결과가 테스트값이 10, 100, 1000, 10000, 100000개 순으로 다음과 같이 나왔어요. (10000, 100000개 일 때는 테스트를 할때마다 1번과 2번의 우세가 바뀌는 의미없는 결과가 나왔어요.)
이렇게 테스트를 하고 나니깐 제가 내린 결론은 "성능적으로는 현 프로젝트에 2번 방법을 적용하는 것이 더 적절해 보인다" 였어요. 하지만 실 현업에서라면(지금은 학습 단계이기 때문에 매우 의미있었던 과정이라고 생각합니다😊) 이러한 성능 속도의 차이가 의미가 있는 것인지. 어떠한 기준으로 구현 방법을 결정하는 지 궁금해지더라구요. 조언해주실 수 있으실까요??
<img width="194" alt="image" src="https://user-images.githubusercontent.com/45311765/154497821-97aa99df-285c-4791-80a5-0f234160c3d4.png">

쓰다보니 질문이 너무 길어졌네요... 😅

heebong replied:
`ParameterizedTest`가 시간이 점점 단축되는게 신기하네요 🤔  `ParameterizedTest`는 말 그대로 해당 테스트를 반복적으로 실행하는건데 캐시처럼 동작해서 빠르게 실행되는건지는 저도 더 찾아봐야겠네요. 새로운 사실을 알려준 엘리 감사합니다 😄 

현업에서는 이정도 속도차이라면 가독성을 우위에 두었을 것 같아요. 코드에서의 정렬이 꼭 필요한 상황이고, 컬렉션의 최대 크기가 테스트한 수라고 보여진다면 우선 가독성에 우위를 둘 것 같습니다. 물론 성능을 조금이라도 개선해야하는 상황이라면 새로운 방법을 찾을 것 같지만요.

heebong commented on src/main/java/racingcar/controller/GameController.java:
```diff
@@ -29,45 +29,52 @@ public void run() {
+        final int tryCount = inputTryCount();
 
-        outputView.printWinners(winners);
+        racingGame = new RacingGame(cars, tryCount);
```
`축약은 어디까지해야하는가..` 질문의 답변 드릴게요.
왜 한 줄로 축약하고 싶으신가요? 왜 한 줄로 만들고 싶을까요? 
물론 함수의 라인 수를 줄이는건 기능별로 분리하기 위함이지만, 그 이유가 아니라면 저는 가독성을 따라갈 것 같아요. 엘리는 어떤게 더 읽기 편한가요? (앗, 이미 질문에 답이 나와있네요 😄 )

RIANAEH replied:
`cars`와 `tryCount`의 값을 `RacingGame` 객체를 생성하는 데 한번만 사용하고 이후 재사용하지 않기때문에 가독성을 위해 변수를 따로 생성해서 값을 할당해주는게 좋을까?에 대해 고민했던것 같아요ㅎㅎ 
조금 더 가독성을 위한 방향으로 진행을 해나가야겠네요 👍

heebong commented on src/test/java/racingcar/model/CarTest.java:
```diff
@@ -15,15 +20,30 @@ void init() {
-        car.moveForward(5);
+    @ParameterizedTest
+    @ValueSource(ints = {4, 5, 6, 7, 8, 9})
+    @DisplayName("자동차 전진 테스트")
```
`@DisplayName 어노테이션의 위치?`에 대한 답을 드릴게요.
제가 알기로 정해진 컨벤션은 없습니다. 다만 저는 맨 밑에 `@DisplayName`을 써요. 엘리는 어디에 쓰는게 더 편한가요? 엘리의 컨벤션을 만들어가면 좋겠네요.

RIANAEH replied:
저도 맨 밑에 `@DisplayName`을 쓰는게 편하더라구요!!
이런 부분에 대해서는 페어 프로그래밍 시 페어와 맞춰나가봐야겠네요ㅎㅎ

heebong reviewed (changes requested):
안녕하세요, 엘리! 리뷰어 희봉입니다 😃 
벌써 자동차 경주 2단계 미션이네요. 이번 미션도 잘 해주셨습니다. 엘리가 더 좋은 코드를 짜기 위해 고민한 흔적이 보여요. 👍  MVC 패턴에 맞게 잘 구현했습니다 💯 

몇 가지 고민점을 코멘트로 남겼습니다. 확인 부탁드려요. 😎  궁금한 내용이 있다면 언제든 슬랙, 혹은 DM 주세요!! 🚗

RIANAEH committed:
refactor(racingcar): 불필요한 테스트 삭제

private 메서드는 public 메서드 테스트로 간접 테스트가 가능하기에 해당 테스트를 삭제하고 테스트를 위해 public으로 변경했던 메소드의 접근 제한자를 다시 private으로 적용했습니다.

RIANAEH committed:
refactor(racingcar): 메서드를 더 세분화

하나의 메서드에서는 하나의 일만 하게 하자!!

RIANAEH commented:
안녕하세요 희봉!!
이번 리뷰 요청에는 실질적인 코드 변경은 거의 없이 질문만 드리게 된 것 같네요 😭
이 다음 요청 떄는 좀 더 성장한 모습을 보여드리겠습니다!!

heebong reviewed (approved):
안녕하세요, 엘리! 리뷰어 희봉입니다 😄  고민도 많이 하고 새로운 시도도 많이 한 것이 느껴지는 요청이었어요. 👍 
이번 과제는 여기서 머지하도록 하겠습니다. 🎉  첫 과제를 진행하시느라 수고하셨어요 🎉   피드백 범위가 아닌 다른 부분까지 바꿔보며 더 나은 코드를 만들어간 엘리를 보며 저도 즐거웠어요. 
앞으로도 과제를 진행하며 많이 배우고 성장하셨으면 좋겠습니다 😎  자동차 경주 완주하느라 수고하셨습니다 🚗 

(추가로 질문에 대한 답은 해당 대화에 댓글로 추가해두었습니다. 궁금한게 있으시면 DM 아시죠? 😉 )
